src/Microsoft.AspNet.Mvc.Core/ModelBinding/MutableObjectModelBinder.cs

```diff
@@ -109,6 +115,13 @@ protected virtual bool CanUpdateProperty([NotNull] ModelMetadata propertyMetadat
 
         private async Task<bool> CanValueBindAnyModelProperties(MutableObjectBinderContext context)
         {
+            // If there are no properties on the model, there is nothing to bind. We are here means this is not a top 
+            // level object. So we return false.
+            if (context.PropertyMetadata == null || context.PropertyMetadata.Count == 0)
```

Comment:
`GetMetadataForProperties` runs user-code and we need to be careful to evaluate the sequence at-most once. This code should have *already* been using something stronger than `IEnumerable`.